### PR TITLE
feat(new sink): Initial `gcp_bigquery` sink implementation

### LINKED ISSUE #1536: New `gcp_big_query` sink
GCP Big Query is a powerful service for analyzing large amounts of structured data. If used correctly, it can be a cost-effective storage for log data. I would like to see Vector support this service as a sink, but it'll require careful planning due to the many different ways Big Query can be used.

## Context

Big Query is flexible and we should consider the following features for our Big Query sink:

1. Storing JSON data in a single column. Users can use Big Query's [JSON functions](https://cloud.google.com/bigquery/docs/reference/standard-sql/json_functions).
    1. This type of querying is slower for obvious reasons (fetching more data, parsing, etc).
    2. This type of querying is more expensive because each query must fetch and scan the entire JSON payload as opposed to individual columns.
2. Mapping structured data to individual Big Query table columns.
3. Automatic schema maintenance.
4. [Streaming inserts](https://cloud.google.com/bigquery/streaming-data-into-bigquery) vs [batching inserts](https://cloud.google.com/bigquery/docs/loading-data).
    1. Streaming inserts [have a cost](https://cloud.google.com/bigquery/pricing) ($0.010 per 200 MB).
    2. Note the variety of methods to batch insert. This can be done through the API directly or through other GCP services (cloud storage, stackdriver, etc).

This, of course, is not inclusive of all factors we should consider for Big Query, but it helps to demonstrate the variety of options we should consider.

## Starting simple

v1 of this sink should solve the simplest implementation:

- [ ] Use the [streaming inserts API]() and stream records 1 by 1 (no batching).
- [ ] Assume we are writing to a table with 2 columns: `timestamp` and `json_event`.
    - [ ] The `timestamp` column should map to our own internal `timestamp` column.
    - [ ] The `json_event` column should contain a JSON encoded representation of our event.
    - [ ] Both of these column names should be configurable. It is worth thinking about a generic column mapping configuration scheme so that users could map other custom fields to Big Query columns.
- [ ] Include documentation on how to create a properly structured table. Ideally this table would be partitioned by the `timestamp` day.

## Long Term

We might consider the following features for long-term development:

1. Support for using the batching API since it does not incur a cost.
2. Dynamic schema maintenance. Although, I think this might be better solved with a transform or something separate.

### PR DESCRIPTION
**Pull request by bruceg**
This sink uses the streaming insert mechanism with batching. No schema rewriting is done. I invite suggestions where that would be best documented.

Ref #1536

### CODE REVIEW

**Comment by binarylogic** [comment]
Since this PR already ingests data directly into columns let's finish that off and merge. We can follow up with an option to place a JSON encoded event into a single column.

File: src/sinks/gcp/bigquery.rs [hunk 1]
```diff
@@ -0,0 +1,193 @@
...
+}
+
+#[derive(Deserialize, Serialize, Debug, Clone, Default)]
+#[serde(deny_unknown_fields)]
+struct BigQueryConfig {
+    project: String,
+    dataset: String,
+    table: String,
+
+    pub emulator_host: Option<String>,
```
Comment (LucioFranco) [inline src/sinks/gcp/bigquery.rs]: I am not 100% following why this is called emulator host? Shouldn't this just be called endpoint that you can optionally point to the emulator url?

**Review by LucioFranco — APPROVED** [review]
LGTM, just a few questions.

File: src/sinks/gcp/bigquery.rs (hunk 1 above)
Comment (bruceg) [reply src/sinks/gcp/bigquery.rs]: The distinction is that, compared to the AWS sources (which have an endpoint) or other sources (that require a hostname), there is *only one* hostname that is valid for this sink, unless you are using an emulator.

And, in fact, even that is wrong, as there is no emulator for BigQuery that could be used for testing, so I think I should scrap even this vestigial setting.

File: src/sinks/gcp/bigquery.rs [hunk 2]
```diff
@@ -0,0 +1,193 @@
...
+    #[serde(flatten)]
+    pub auth: GcpAuthConfig,
+
+    #[serde(default)]
+    pub batch: BatchBytesConfig,
+    #[serde(default)]
+    pub request: TowerRequestConfig,
+
+    pub tls: Option<TlsOptions>,
```
[the remark this answers is not in the document]
Comment (bruceg) [reply src/sinks/gcp/bigquery.rs]: TLS is required for GCP services, so there is no point including the `enabled` option that is part of `TlsConfig`.

**Comment by bruceg** [comment]
Ben, what needs to be said about schema conformity here?

**Comment by binarylogic** [comment]
Let's chat about this so that I can confirm the behavior. Then I get help write up the docs.

**Comment by Hoverbear** [comment]
@binarylogic @bruceg did you two get a chance to chat?

**Comment by Hoverbear** [comment]
@bruceg Can you help me understand the status here? Should this be re-reviewed?

**Comment by binarylogic** [comment]
This is entirely blocked by me. I wanted to update the docs. This sink is our first sink that requires a strict schema, so I wanted to make sure our docs educate users on how to enforce and align schemas with their Big Query table.

**Comment by Hoverbear** [comment]
Removing Bruce and assigning you then. Thanks so much for this great feature @bruceg  :)

**Comment by jamtur01** [comment]
@binarylogic This one is your inbox.

**Comment by jamtur01** [comment]
Closing for now. @binarylogic please re-open if you get a chance to update the docs.

**Comment by binarylogic** [comment]
This PR was originally blocked by the docs because it required a strict schema to use properly, but we've decided to proceed anyway, hence why it's been in @bruceg's sprint. At this point it is not blocked by docs, just a rebase with master, assuming it has not drifted too far.

**Comment by binarylogic** [comment]
For the conflicts, I just need @bruceg to resolve:

```
src/sinks/gcp/cloud_storage.rs
src/sinks/gcp/mod.rs
src/sinks/gcp/stackdriver_logs.rs 
```

The rest can be ignored and overwritten with what's in `master`.

**Comment by binarylogic** [comment]
Given the age of this PR I'm going to close it. We can revive it when we get to implementing this sink.

**Pull request closed without merging**